Save Discogs Release ID via MusicBrainz

JOJ0 opened the pull request:
## Description

- Fixes #4220 
- Enable fetching 'url-rels' from the musicbrainz release endpoint.
- A musicbrainz release might contain a link to a discogs release or a discogs master release.
- We prefer the regular release and save it to the beets library field discogs_albumid. as a fallback we save the master release id.
- For extraction of the release id from the discogs url, we use the methodfound in the discogs plugin.
- Todo: The discogs plugin needs to be refactored to use the method outsideof the DiscogsPlugin class...

## To Do

- [ ] Documentation. (If you've add a new command-line flag, for example, find the appropriate page under `docs/` to describe it.)
- [ ] Changelog. (Add an entry to `docs/changelog.rst` near the top of the document.)
- [ ] Tests. (Encouraged but not strictly required.)

JOJ0 commented:
Is it ok if I move the extract_release_id_regex method out of the DiscogsPlugin class but leave it in the module so others can import it from there? Or should I rather move it into the general beets.util module?

JOJ0 commented:
> Is it ok if I move the extract_release_id_regex method out of the DiscogsPlugin class but leave it in the module so others can import it from there? Or should I rather move it into the general beets.util module?

Doesn't work like that! I was getting a circular import error with that approach thus I decided that moving to beets.util for the discogs extractor function is possible option: https://github.com/beetbox/beets/pull/4474/commits/ce5d659a9f8e924c32ca35f4f86994a59c922c47

Dear maintainers, can you think of a better place or approach to make this helper method usable from within the discogs plugin but also from other plugins, or if my solution is ok like that. Thanks in advance!

JOJ0 commented:
@wisp3rwind since you were involved in the process of extending the discogs url to id extractor method a couple of months ago. I extended it with the short form url format that is used very often in the discogs relation links saved in musicbrainz releases (and also my go to move if I have a discogs release id at hand and want to quickly look it up). May I ask for an early review from you: I extended the regex and saved a new version on regex101: https://github.com/beetbox/beets/pull/4474/commits/693413146560fa7498e913e17a83a42f23e961be

arsaboo commented:
This is pretty epic. I think it would be great to have this function accessible in other plugins. For example, I would love to be able to translate between MusicBrainz artist_id and Spotify artist_id.

JOJ0 commented on beets/autotag/mb.py:
```diff
@@ -466,6 +471,43 @@ def album_info(release):
+        if discogs_url:
+            info.discogs_albumid = util.extract_discogs_id_regex(discogs_url)
+        if spotify_url:
+            info.spotify_albumid = "FIXME id extractor method?"
```
Hi @arsaboo, before I dig into the Spotify plugin code, I'm sure you know by heart! :-) Is there an ID extractor method in the plugin similar to the extrac_discogs_id_regex() method? Thanks in advance!

arsaboo replied:
Hey @JOJ0....this is super cool. Being able to get all the details from MB would be awesome. 

The Spotify plugin uses the `get_id` method to obtain the id (and that method is defined in the plugins.py):
https://github.com/beetbox/beets/blob/b11df49705b877e3753c165a444f9529fe59f19a/beetsplug/spotify.py#L333

You should be able to use that method. BTW, I am assuming that you are checking if `spotify_albumid` is already populated before overwriting that field.

JOJ0 replied:
I will try to use that function from outside the plugin. Not sure if it will work, had massive problems with that with the discogs_id_extractor method. How can I import the this `_get_id()` method from the plugin do you think? I'll try but with the discogs_id_extractor, "I got circular import errors" and in the end had to move that function out of the plugin to make use of it.

JOJ0 replied:
And thanks for the reminder that checking if it is already populated! But how would I know if it is correct or the user wants to overwrite it when reimporting?? There is not really I way to know what the user wants other than believing what he configured: If he set the new config directive to yes, then it will be overwritten. At least that would have been my plan. What do you think?

```
musicbrainz:
    url_rels:
        spotify: yes
```
-> would mean: I want to overwrite it!

Also brainstorming here: "url_rels" in the config is not very self-explanatory (it's called like that in the muscibrainz json response). Do you have a better idea on how this config directive should be called? Thanks in advance!

arsaboo replied:
Spotify refers to them as `external_ids` - basically all external IDs (besides MB can be part of it, including ISRC and anything else that may be available). Maybe we can use the same terminology here and explain more in the docs. 

The default behavior should be no-overwrite. Currently, the `spotify_albumid` is only populated when the user has explicitly used the Spotify tagger and manually matched it. Thus, it does not make sense to overwrite this information unless the user explicitly requests it. I think we should use a flag `-f` (or a config entry) to force overwrite if the field is already populated. The Spotify plugin uses a similar flag. 

I did not write that part, but for the `_get_id()`, you will have to import `MetadataSourcePlugin`
https://github.com/beetbox/beets/blob/b11df49705b877e3753c165a444f9529fe59f19a/beetsplug/spotify.py#L34

JOJ0 replied:
Thanks so much for your hints @arsaboo!

Using MetadataSourcePlugin._get_id method doesnt work like that:

```
  File "/home/jojo/git/beets/beets/autotag/mb.py", line 510, in album_info
    info.spotify_albumid = MetadataSourcePlugin._get_id('album', 123)
TypeError: _get_id() missing 1 required positional argument: 'id_'
```

When looking at the method, it's obvious. It is contained in an abstract class and the variable self._regex needs to be implemented in the source plugin that is based on `MetadataSourcePlugin: https://github.com/beetbox/beets/blob/b11df49705b877e3753c165a444f9529fe59f19a/beets/plugins.py#L706

So I can't just import from MetaDataSourcePlugin and expect things to work. This is what I'm missing from the spotify plugin: https://github.com/beetbox/beets/blob/b11df49705b877e3753c165a444f9529fe59f19a/beetsplug/spotify.py#L74

JOJ0 replied:
So the next obvious step would be to try importing from the spotify beetsplug directly. Now I'm hitting exactly the same error as I had with the discogs plugin. circular import error. I think using stuff from a beets plugin outside of that plugin, is just not supported as it is implemented right now.

```
  File "/home/jojo/git/beets/beets/autotag/mb.py", line 29, in <module>
    from beetsplug import spotify
  File "/home/jojo/git/beets/beetsplug/spotify.py", line 32, in <module>
    from beets.autotag.hooks import AlbumInfo, TrackInfo
ImportError: cannot import name 'AlbumInfo' from partially initialized module 'beets.autotag.hooks' (most likely due to a circular import) (/home/jojo/git/beets/beets/autotag/hooks.py)
```

I don't really now how to proceed here. BTW, if you have a look at my previous commits, this was the reason I had to move the discogs_id_extractor method to the utils module....I'm still not sure if this is they way to go though....

arsaboo replied:
I think the difference between Spotify and the current PR is that Spotify is a plugin (inside `beetsplug`), whereas you are modifying the `mb.py` directly. I am just thinking out aloud, but wondering if it will be easier to just create another plugin (say `mbextend`) to add these features.

JOJ0 replied:
Actually I thought that mb.py is a plugin by itself. I thought I recall a PR, where musicbrainz was refactored to be a source plugin on it's own like all the others? Did I dream that? Or was it never finished? Will investigte PR's....

Ok about your thought: If I create beetsplug/mbextend.py, don't you think that I will run in exactly the same circular dependency error again?

Maybe I could try if it theoretically would work to run a method of one plugin in another plugin like this: I use discogs as the source plugin, and then try importing with `from beetsplug import spotify`

arsaboo replied:
actually, you may not even need to import Spotify. The `get_id` function is part of `MetadataSourcePlugin` which can be directly imported.

JOJ0 replied:
This is what I tried to tell you in this post: https://github.com/beetbox/beets/pull/4474#discussion_r980796263

It can’t work! MetadataSourcePlugin is missing self.id_regex. It’s not existing. This is to be implemented in the child class, in this case the SpotifyPlugin class. Have a look in MetadataSourcePlugin code!

arsaboo replied:
Exactly, and I am thinking that you can create a new `MBExtendPlugin` class in the new plugin that achieves the same. Maybe I am missing something.

JOJ0 replied:
Yeah I think you miss the fact that it would be clumsy to having to reimplement the regex again in that new plugin. Double code fir the same thing! Dont you think it would be better if this regex or the id extractor function as a whole is moved to a place outside of the spotify plugin, so others can use it?

- Have a look at what I did wit the discogs regex function in my previous commits. To make it usable from outside the discogs plugin I had to move it away from the plugin. I chose the utils module. Now I can use it.

similar provlem with spotify regex. should I move it outside of the spotify plugin?? Only way I see currently if we dont want to copy code!

you see the problem now?

sampsyo reviewed (commented):
Looks good overall so far! I like the idea of moving the Discogs ID extractor utility to `util` to avoid duplication. :+1:

arsaboo commented on beets/config_default.yaml:
```diff
@@ -118,6 +118,10 @@ musicbrainz:
     genres: no
+    url_rels:
+        discogs: no
+        spotify: no
```
I think we should have an `overwrite` flag as well. If a user has already matched using one of the other trackers, say `spotify`, we should not overwrite that information unless the user explicitly asks for that.

JOJ0 replied:
it's in the works ;-)

JOJ0 replied:
Alright, I'm not really sure what I'm doing here....does this sound like the right place at all to implement an overwrite flag: https://github.com/beetbox/beets/pull/4474/commits/8f864b44f2e70c79fb7b9fc4a70bbfb5b0245547

JOJ0 replied:
To be honest, I probably should bother @sampsyo here: With above's commit, why do I get an item that for example has an empty discos_albumid field even though beet info -l shows that it actually is populated:

Debug print of the item inside the apply_metadata function.
```
discogs_albumid=0
```

but beet info is showing that it's populated in the db:

```
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet info makoto golden -l | grep discogs_album
     discogs_albumid: 389624
     discogs_albumid: 389624
(beets)    ~/git/beets   mb_fetch_discogs_url ●  
```

sampsyo replied:
The most likely culprit is that this information is present only at the album level, not the track level (or vice versa). Is there a chance that the field is only applied to one or the other?

JOJ0 replied:
Hi! Thanks so much for the quick response and putting my in the right direction! Funnily enough yesterday evening I did another poking around debugging session and found out exactly that! Yes indeed this field is only set on the album level and not on the items!

JOJ0 replied:
Ok I answered to quickly. Again I confused the album_info with the mapping. I still don't know how to get the "current discogs_albumid set in the beets database already" in this place in the code. I think it is just not possible here and I need to look somewhere else.... Looking into the database directly I find the information saved in the album table but as well on each track in the items table!

Spelling it out loudly:

album_info is the new information we just fetched and set in autottag.mb.album_info() -> We only have new information in here and can't compare with old information!

mapping seems to be, at least to my (probably wrong) understanding, the current information we have in the database already in the form of TrackInfo objects. It does not contain the discogs_albumid.

JOJ0 replied:
After around seven hundred debug messages all around beets code I found out that:

When reimporting via the file path:
```
beet -v import ~/Sync/beets_test_music/Good\ Looking\ -\ Makoto\ Featuring\ MC\ Conrad
```
the discogs_albumid attribute is not set on the TrackInfo object because on my test file it is not existing as a file tag but  _only_ set in the beets db!

When reimporting from file the construction of an Item object seems to work a little different. Trying to understand what exactly is happening here at the moment.... https://github.com/beetbox/beets/blob/master/beets/library.py#L602-L604

On the other hand, when reimporting via a library query:

```
beet -v import -L makoto golden
```

my code submitted in https://github.com/beetbox/beets/commit/8f864b44f2e70c79fb7b9fc4a70bbfb5b0245547 works perfectly fine: The Trackinfo object _does_ have the discogs_albumid property set! (Even though it is _only_ available as a database field and _not_ available from the media file's tag)

So actually what I need to find out currently, is:
- Why is discogs_albumid not written to the media file when it is populated with this PR's new feature?
- Is it expected behaviour that an Item object when reimported via the file path, does not contain attributes that are only set in the db, but are not available from the file's tags.

JOJ0 replied:
> * Why is `discogs_albumid` not written to the media file when it is populated with this PR's new feature?

The answer is obvious, I now recall how writing to media files works in beets: discogs_albumid is never written to disk, since the Discogs plugin does not add the MediaField. This could be made possible with something like that in the plugin's initialization (but that's probably something for another PR, just saving the thougth):

```
        self.add_media_field("discogs_albumid", MediaField(
            MP3DescStorageStyle(u'discogs_albumid'),
            StorageStyle(u'discogs_albumid')
        )
        # and further storage styles for MP4 and ASF
```

Also I realize that none of the metadatasource plugins do add media_fields, thus with the current state of beets we never get file tags `spotify_album_id` `deezer_album_id`, `beetport_album_id`, and so on. Also there currently is no way of configuring that these fields should be written to disk. Correct my if I'm wrong!

JOJ0 replied:
Ok, I'm about to give up. I don't understand why this is not working and it's getting pretty frustrating by now to not being able to achieve the simple task of storing something in a database. There must be something obvious I'm not seeing.

So please @sampsyo, @arogl , @arsaboo, anyone, ... can you have a look at this and tell me if by any means this could work: https://github.com/beetbox/beets/pull/4474/commits/7206083d6864d6f1109ec2c679d93151d7fc9590

The Initial import of the id fields via MusicBrainz works fine but as soon as I try to reimport with the external_ids_overwrite flag set to true, I get into trouble. It just doesn't seem to be possible to update the values in the database. I thought I had it, thought it goes wrong in this place because it overwrites the new values with values from the current items in the database but it doesn't do the trick.

So for example let's take this 2-track single:
```
$ beet -v import ~/Sync/beets_test_music/Robert\ Hood\ -\ Eleven\ \[M.PM18\]\ \(2013\)\ ALAC\ mp3tag/
...
...
$ beet info -l robert hood alarm | grep spotify
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV

$ beet -v edit robert hood alarm
# change one of the ID's to something else

$beet info -l robert hood alarm | grep spotif
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: ABCD

$ beet -v import ~/Sync/beets_test_music/Robert\ Hood\ -\ Eleven\ \[M.PM18\]\ \(2013\)\ ALAC\ mp3tag/ 
...
Reimported album: added 1666681278.518687, flexible attributes [] from album 31 for /home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag
Reimported item added 1666681278.5210571 from item 75 for /home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag/01 - Robert Hood - Eleven - Gm,Abm.m4a
Reimported item flexible attributes ['spotify_album_id', 'track_alt', 'data_source', 'track_alt', 'spotify_album_id', 'data_source'] from item 75 for /home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag/01 - Robert Hood - Eleven - Gm,Abm.m4a
Sending event: database_change
Reimported item added 1666681278.524402 from item 76 for /home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag/02 - Robert Hood - Alarm - Am.m4a
Reimported item flexible attributes ['spotify_album_id', 'track_alt', 'data_source', 'track_alt', 'data_source', 'spotify_album_id'] from item 76 for /home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag/02 - Robert Hood - Alarm - Am.m4a

...

$ beet info -l robert hood alarm | grep spotif                                                        
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: ABCD
```

arsaboo replied:
Can you try breaking up this part and adding debug logs to see if we have the right flow?

https://github.com/beetbox/beets/blob/7206083d6864d6f1109ec2c679d93151d7fc9590/beets/autotag/__init__.py#L195-L197

Specifically, create two if conditions for `if not item.get('spotify_album_id') or ext_id_force:`. Since the first import worked, I am assuming that the `not item.get('spotify_album_id')` is working and the tag writing is working fine.

JOJ0 replied:
Hi @arsaboo, thanks so much for helping with debugging! I checked this condition a 100 times already. It is definitely working as expected and I'm pretty sure the problem lies somewhere else! Anyway, as suggested I torned apart the condition into two branches and put some debug print messages. This is the commit: 7f3a914050675244312fe1c98e4aa49bd15ad061, and this is the output

```
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet rm robert hood alarm 
Robert Hood - Eleven / Alarm - Eleven
Robert Hood - Eleven / Alarm - Alarm
Really remove 2 items from the library?? (Yes/no/select) y

(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet import ~/Sync/beets_test_music/Robert\ Hood\ -\ Eleven\ \[M.PM18\]\ \(2013\)\ ALAC\ mp3tag/  

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
WRITING spotify_album_id because it's empty yet.
id was:  None
id will be:  2DWmMUYp5LQNFu1aulfmHV
WRITING spotify_album_id because it's empty yet.
id was:  None
id will be:  2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet import ~/Sync/beets_test_music/Robert\ Hood\ -\ Eleven\ \[M.PM18\]\ \(2013\)\ ALAC\ mp3tag/ 

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
WRITING spotify_album_id because it's empty yet.
id was:  None
id will be:  2DWmMUYp5LQNFu1aulfmHV
WRITING spotify_album_id because it's empty yet.
id was:  None
id will be:  2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet import -L robert hood alarm                                                                 

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  2DWmMUYp5LQNFu1aulfmHV
id will be:  2DWmMUYp5LQNFu1aulfmHV
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  2DWmMUYp5LQNFu1aulfmHV
id will be:  2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url ●  

```

JOJ0 replied:
Ok let's talk about problem number 2: Here first without any new debug messages, just to face the problem:

```
(beets)    ~/git/beets   mb_fetch_discogs_url  beet info robert hood alarm | grep spotify                  
(beets)      ~/git/beets   mb_fetch_discogs_url  
(beets)      ~/git/beets   mb_fetch_discogs_url  beet info -l robert hood alarm | grep spotify
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url  beet edit robert hood alarm               
Robert Hood - Eleven / Alarm - Eleven
  spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV -> ABCD
continue [E]diting, Apply, Cancel? a
(beets)    ~/git/beets   mb_fetch_discogs_url  beet info -l robert hood alarm | grep spotify
    spotify_album_id: ABCD
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url  beet import -L robert hood alarm             

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  ABCD
id will be:  2DWmMUYp5LQNFu1aulfmHV
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  2DWmMUYp5LQNFu1aulfmHV
id will be:  2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url  
(beets)    ~/git/beets   mb_fetch_discogs_url  beet info -l robert hood alarm | grep spotify
    spotify_album_id: ABCD
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url  

```

JOJ0 replied:
After a lot of debugging I found out that the problem might be in 'importer.ImportTask.reimport_metadata`, thus let's have a look at this debug commit 7206083d6864d6f1109ec2c679d93151d7fc9590 and it's output:


```
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet info robert hood alarm | grep spotify   
(beets)      ~/git/beets   mb_fetch_discogs_url ●  beet info -l robert hood alarm | grep spotify
    spotify_album_id: ABCD
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet import -L robert hood alarm             

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  ABCD
id will be:  2DWmMUYp5LQNFu1aulfmHV
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  2DWmMUYp5LQNFu1aulfmHV
id will be:  2DWmMUYp5LQNFu1aulfmHV
We are in importer.ImportTask.reimport_metadata:
dup_item.spotify_album_id is non-empty
Value of item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
Value of dup_item.spotify_album_id is: ABCD
We've now set item.spotify_album_id to: ABCD
We've also added spotify_album_id to item._dirty: {'discogs_albumid', 'added', 'spotify_album_id'}
dup_item.spotify_album_id is non-empty
Value of item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
Value of dup_item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
We've now set item.spotify_album_id to: 2DWmMUYp5LQNFu1aulfmHV
We've also added spotify_album_id to item._dirty: {'discogs_albumid', 'added', 'spotify_album_id'}
(beets)    ~/git/beets   mb_fetch_discogs_url ●  
(beets)    ~/git/beets   mb_fetch_discogs_url ●  beet info -l robert hood alarm | grep spotify
    spotify_album_id: ABCD
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
(beets)    ~/git/beets   mb_fetch_discogs_url ●  

```

JOJ0 replied:
@arsaboo I finally realized what my error was. I had to change dup_item instead of item in my last commit. This fixes it: 0f16e2a7483159c1380a1ba47d03c5e1886aac10

Here's the output:

```
(beets)    ~/git/beets   mb_fetch_discogs_url  beet info -l robert hood alarm | grep spotify                             
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: ABCD
(beets)    ~/git/beets   mb_fetch_discogs_url  beet import -L robert hood alarm                                          

/home/jojo/Sync/beets_test_music/Robert Hood - Eleven [M.PM18] (2013) ALAC mp3tag (2 items)
Tagging:
    Robert Hood - Eleven / Alarm
URL:
    https://musicbrainz.org/release/79ef4094-55db-4fbb-889a-7807ab692494
(Similarity: 100.0%) (Digital Media, 2013, US, M-Plant, M.PM18)
 * Eleven
 * Alarm
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  2DWmMUYp5LQNFu1aulfmHV
id will be:  2DWmMUYp5LQNFu1aulfmHV
OVERWRITING spotify_album_id because forced overwrite is enabled.
id was:  ABCD
id will be:  2DWmMUYp5LQNFu1aulfmHV
We are in importer.ImportTask.reimport_metadata:
item.spotify_album_id is non-empty
Value of item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
Value of dup_item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
We've now set dup_item.spotify_album_id to: 2DWmMUYp5LQNFu1aulfmHV
We've also added spotify_album_id to item._dirty: {'added', 'spotify_album_id', 'discogs_albumid'}
item.spotify_album_id is non-empty
Value of item.spotify_album_id is: 2DWmMUYp5LQNFu1aulfmHV
Value of dup_item.spotify_album_id is: ABCD
We've now set dup_item.spotify_album_id to: 2DWmMUYp5LQNFu1aulfmHV
We've also added spotify_album_id to item._dirty: {'added', 'spotify_album_id', 'discogs_albumid'}
(beets)    ~/git/beets   mb_fetch_discogs_url  
```

arsaboo replied:
That is some amazing work. Kudos for not giving up and continuing to address the issue. 

Hopefully, the end is in sight. I am so excited to see this feature.

arogl replied:
@JOJ0 I also noticed that there were 2 spotify_album_id in the output. Is this normal or should one of them be a track_id like MB and discogs?

``` sh
(beets)    ~/git/beets   mb_fetch_discogs_url  beet info -l robert hood alarm | grep spotify                             
    spotify_album_id: 2DWmMUYp5LQNFu1aulfmHV
    spotify_album_id: ABCD
(beets)    ~/git/beets   mb_fetch_discogs_url  beet import -L robert hood alarm                                          
```

JOJ0 replied:
@arogl I'm not sure if I understand the concern. You migth have missed that I'm testing with 2 media files and I'm changing one of them to ABCD. The other one I didnt touch. But you might be talking about something else

arogl replied:
@JOJ0 

All OK, I thought it was only 1 file with a duplicated tag.

I looked at the link and see it is a 2-track album.

Well done on the debugging

arogl commented on beets/autotag/mb.py:
```diff
@@ -475,6 +478,59 @@ def album_info(release):
+    # FIXME This check is always True because of config_default.yaml.
+    if ('url_rels' in config['musicbrainz'].keys() and
+            release.get('url-relation-list')):
+        discogs_url, deezer_url = None, None
+        spotify_url, bandcamp_url, beatport_url = None, None, None
+        fetch_discogs, fetch_spotify, fetch_bandcamp = False, False, False
+        fetch_beatport = False
```
@JOJ0 These look like they need to be set above the if statement according to the failed report.

JOJ0 replied:
Oh! Thanks a lot! Fixed in the original commit.

arogl commented on beets/autotag/mb.py:
```diff
@@ -475,6 +478,59 @@ def album_info(release):
+        discogs_url, deezer_url = None, None
+        spotify_url, bandcamp_url, beatport_url = None, None, None
+        fetch_discogs, fetch_spotify, fetch_bandcamp = False, False, False
+        fetch_beatport, fetch_deezer = False
```
@JOJ0 needs ", False"

JOJ0 replied:
Seems to be my "Thanks" a couple of days ago didn't get through. So again: Thanks a lot! Also fixed in original commit! :-)

JOJ0 commented:
Hi @sampsyo as recently discussed in #4537 I removed the "prevent ID overwrite" functionality and finished this PR. I'd kindly request a final review. I try to summarize what I think should be looked into in particular:

- As you've already approved, I moved the Discogs id extraction to the `utils` module.
- Some parts, in particular the actual regex strings of all the available metadata source plugins, had to move there as well.
- I went a step further and even invented a submodule in utils named `id_extractors` because I think it's much cleaner than putting more stuff into the already poluted utils module directly: https://github.com/beetbox/beets/pull/4474/files#diff-52a89c1045582fbed2d44321767c31a4b6e70c836ac234e60b2fcb26f64c4702
- The id extraction method in the `MetadataSourcePlugin` abstract class had to be refactored just a little. It is now a staticmethod and it's possible to use it from the outside of plugins: https://github.com/beetbox/beets/pull/4474/files#diff-165009439d41875b686d8cd7d4a1d698bcd33d53bf2d8748d13f3e9c59592c33
- The metadata source plugins required tiny refactoring as well: https://github.com/beetbox/beets/pull/4474/files#diff-165009439d41875b686d8cd7d4a1d698bcd33d53bf2d8748d13f3e9c59592c33.
- The beatport plugin had some things built-in that's actually covered by the abstract class already. I refactored it, it's now streamlined with the other plugins that are using the abstract class.


- I decided to always include the additional url-rels data when musicbrainz is requested in the autotagger because I _suppose_ it's not much of a performance impact - open for discussion if you think it's not a good idea: https://github.com/beetbox/beets/pull/4474/files#diff-ff51bca3f15a32bf8652a69e127b582652187bae2fd2fd7a1ccb58d0feeb97b2
- The `reimport_metadata` method in the importer required some "exclusion code" to prevent the overwriting of flexible  attributes on reimports. I left a longish comment there to make it clear why this is there. https://github.com/beetbox/beets/pull/4474/files#diff-bb5be0a837ad397e52d11fc49be3bbbff18df98a6feae43208dd2231929f2fed


I hope that helps with reviewing in particular to understand the _why_'s of some of my changes quicker!

JOJ0 commented:
I was wondering whether it would help maintainers (in that most probably this is directed to @wisp3rwind  and @sampsyo) if I split up this PR into 2 pieces? Human resources in the project are still rare I guess and I played with the thought that it might help to first submit a PR implementing _only_ the "refactoring metadata ID extraction" part and then the actual feature. I think the former is a nice addition to beets on it. It even fixes and streamlines things and I think that adds to the project in its own way. And certainly my hope is that after that would be merged, the actual feature I'd like to bring into beets would result in a slightly smaller and hopefully a little easier to review PR.

What do you think? Would that make sense? I have a few days off from work and could start working on it soon.

wisp3rwind commented:
> I was wondering whether it would help maintainers (in that case most probably this is directed to @wisp3rwind and @sampsyo) if I split up this PR into 2 pieces? Human resources in the project are still rare I guess and I played with the thought that it might help to first submit a PR implementing _only_ the "refactoring metadata ID extraction" part and then the actual feature. I think the former is a nice addition to beets on its own. It even fixes some things and most of all streamlines things and I think that adds to the project already. And certainly my hope is that after that would be merged, the actual feature I'd like to bring into beets would result in a slightly smaller and hopefully a little easier to review PR.
> 
> What do you think? Would that make sense? I have a few days off from work and could start working on it soon.

Splitting things into small orthogonal PRs sounds like a good idea! (Even though, at first sight, this PR doesn't appear exactly gigantic.) That being said, don't rely on me personally reviewing this very soon: There's limited time I can and want to spend on beets; and one thing that I've been trying to do is focus reviews on parts of beets that I actually know about. The metadata sources are decidedly not in that scope, and there are already a few other important PRs that I really should get around to (for which I've been involved with already, in contrast to this PR); sorry!

JOJ0 commented:
Thanks a lot for an honest answer, letting me know what to expect from your angle. Appreciate it! Since I use this feature everyday anyway from my own branch I don't have any rush in getting this reviewed. I'll consider creating a PR-split but at the moment I picked up other tasks related to my beets tinkering and config.  Thanks for letting me know that you think it is basically a good idea.

JOJ0 commented:
Changed this PR back to draft and "split out" the metadata ID extraction part into a new PR: #4663

JOJ0 commented:
Closing in favor of new PR #4220
